sled-agent/src/storage_manager.rs

```diff
@@ -863,13 +880,151 @@ impl StorageManager {
             underlay_address,
         };
         StorageManager {
+            log,
+            disks,
             pools,
             new_pools_tx,
             new_filesystems_tx,
             task: tokio::task::spawn(async move { worker.do_work().await }),
         }
     }
 
+    /// Ensures that the storage manager tracks exactly the provided disks.
+    ///
+    /// This acts similar to a batch [Self::upsert_disk] for all new disks, and
+    /// [Self::delete_disk] for all removed disks.
+    pub async fn ensure_using_exactly_these_disks<I>(
+        &self,
+        unparsed_disks: I,
+    ) -> Result<(), Error>
+    where
+        I: IntoIterator<Item = crate::hardware::UnparsedDisk>,
+    {
+        let mut new_disks = HashMap::new();
+
+        // We may encounter errors while parsing any of the disks; keep track of
+        // any errors that occur and return any of them if something goes wrong.
+        //
+        // That being said, we should not prevent access to the other disks if
+        // only one failure occurs.
+        let mut err: Option<Error> = None;
+
+        // Ensure all disks conform to the expected partition layout.
+        for disk in unparsed_disks.into_iter() {
+            match crate::hardware::Disk::new(&self.log, disk).map_err(|err| {
+                warn!(self.log, "Could not ensure partitions: {err}");
+                err
+            }) {
+                Ok(disk) => {
+                    let devfs_path = disk.devfs_path().clone();
+                    new_disks.insert(devfs_path, disk);
+                }
+                Err(e) => {
+                    warn!(self.log, "Cannot parse disk: {e}");
+                    err = Some(e.into());
+                }
+            };
+        }
+
+        let mut disks = self.disks.lock().await;
+
+        // Remove disks that don't appear in the "new_disks" set.
+        //
+        // This also accounts for zpools and notifies Nexus.
+        let keys_to_be_removed = disks
+            .keys()
+            .filter_map(|key| {
+                if !new_disks.contains_key(key) {
+                    Some(key.clone())
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>();
```

Comment:
I updated to mostly use this format, with one wrinkle. I agree that the problem area of "disk was re-inserted, but with new data" is... weird. It's complex. I dunno how to arbitrarily handle it, especially if subtle aspects of the disk somehow change.

I've opted to treat this as a case of "remove the disk", and then "add the disk". Open to suggestions on what else to do, but it would certainly be unexpected if a disk started reporting a new partition layout / zpool / slot while we were in the middle of using it.